### PR TITLE
Add use_accuracy_compatible mode for MoE/MLP numerical alignment

### PR DESCRIPTION
**Pull request by zhengshengning**
### PR Category

Operator Mechanism 

### PR Types

Bug fixes

### Description

**概述**
1. 引入 use_accuracy_compatible 开关，贯穿 MLP、MoE 层、router、token dispatcher 和 SwiGLU 融合，使数值行为与参考实现对齐。

**改动**
1. fused_bias_swiglu：新增 eager 版 SwiGLU 前向/反向路径（swiglu_eager、bias_swiglu_eager、swiglu_back_eager、weighted_swiglu_back_eager），由 use_accuracy_compatible 控制。
2. tensor_parallel/layers：linear 反向中对 expert 参数使用 fp32 main_grad 累加；为权重打上 is_expert_param 标记。
moe_layer：新增 ThreePathCloneAlignMG clone，其反向按 MG 对齐的求和顺序执行；router / dispatcher / shared 输入分别走独立路径。
3. moe_router：aux-loss 重算、gather_nd 方式的 top-k 权重 gather、以及 fp64 归一化分母对齐。
moe_utils / token_dispatcher：与 MG 对齐的 gather/sum 版 permute/unpermute PyLayer，并接入各个 dispatcher。

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 1]
```diff
@@ -1248,10 +1271,22 @@ def forward(
...
+            _hs_router_path, _hs_dispatcher_path, _hs_shared_path = (
+                ThreePathCloneAlignMG.apply(hidden_states)
+            )
+            residuals = _hs_shared_path
         _log_moe_md5(hidden_states, "moe_input", layer_idx)
 
         self._maybe_pre_allgather_overlap(hidden_states)
         gate_input = self._prepare_gate_input(hidden_states, residual)
+        if _three_paths_enabled:
+            gate_input = _hs_router_path
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_layer.py]: P1 这里在调用 `_prepare_gate_input` 后又把结果替换成了 `hidden_states` 的 clone，下面的 expert 路径也同样覆盖了 `_prepare_expert_input`。`Gemma4MoELayer` 明确使用 `residual` 做 router 输入，并对 expert 输入执行 `pre_feedforward_layernorm_2`；开启该模式训练时会静默绕过这两个 hook，改变模型拓扑。请避免三路径逻辑覆盖 hook 的结果（或在双输入子类上禁用它），并补充 `use_accuracy_compatible=True` 且传入 `residual` 的回归测试。

File: src/paddlefleet/transformer/moe/moe_utils.py [hunk 2]
```diff
@@ -154,6 +154,97 @@ def _unpermute_fp32_accum(
...
+    global_position = position_in_expert_T + expert_offsets[:-1].unsqueeze(
+        1
+    )  # [num_experts, num_tokens]
+    global_position_per_token = global_position.T  # [num_tokens, num_experts]
+
+    topk = int(routing_map_bool.cast("int64").sum(axis=-1)[0].item())
+    valid_positions = global_position_per_token * routing_map_bool.cast("int64")
+    reverse_indices_flat = paddle.masked_select(
+        valid_positions, routing_map_bool
+    ).reshape([num_tokens * topk])
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_utils.py]: P1 Router 会把 padding token 的 routing row 清零，因此混合有效 token/padding 的 map 是合法输入。此时 `reverse_indices_flat` 只有 `有效 token 数 * topk` 个元素，这里却按 `总 token 数 * topk` reshape；若首行恰好是 padding，上一行还会把 `topk` 读成 0。结果是 aligned combine 在常见的 padded batch 上直接 shape error，`_PermuteAlignedPyLayer.backward` 也有同样的全量 reshape 假设。请按每行有效路由数恢复 padding 行的零输出/零梯度，并增加“有效行与全零行混合”的前反向测试。

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 3]
```diff
@@ -620,16 +643,29 @@ def _cal_seq_aux_loss(
...
+            if self.use_accuracy_compatible:
+                tokens_per_expert = routing_map.sum(
+                    axis=seq_axis, dtype="float32"
+                )  # [B, E]
+                _aggregated = all_probs.sum(axis=seq_axis)  # [B, E]
+                _per_expert = _aggregated * tokens_per_expert  # [B, E]
+                _scalar = float(self.num_experts) / (
+                    float(top_k) * float(max_seq_len) * float(max_seq_len)
+                )
+                seq_aux_loss = _per_expert.sum() * _scalar
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_router.py:669]: P1 这里对 `[B, E]` 直接求和，却只按 `S²` 缩放，缺少 sequence-level aux loss 所需的 batch 平均；例如无 padding 且 `B=2` 时，loss 会恰好放大 2 倍。另外前面已计算的每行有效 token `denom` 在该分支完全未使用，padding 序列仍按固定 `max_seq_len²` 归一化。请按 batch 求均值并使用有效序列长度；新增测试应覆盖 `B>1` 和每行 padding 长度不同的情况，而不是把当前求和公式本身作为 expected。

File: src/paddlefleet/transformer/moe/moe_router.py [hunk 4]
```diff
@@ -1331,7 +1375,14 @@ def forward(self, input, input_ids=None, origin_input_ids=None):
         _log_moe_md5(gates, "gate_probs_sigmoid", self._layer_number)
 
         # Use clone() to ensure that the execution order of the grad nodes is consistent with EC.
-        gates_ori = gates.clone()
+        if self.use_accuracy_compatible:
+            gates_ori = paddle.nn.functional.sigmoid(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_router.py]: P1 `use_accuracy_compatible` 并未限制 `scoring_func`，且默认值是 `softmax`，但这里把 aux-loss 使用的 `gates_ori` 无条件重算成了 sigmoid。`seq_aux_loss` 随后会基于它重建 top-k map 并回传梯度，因此 softmax router 开启该开关后训练目标被改成了 sigmoid。请通过配置的 `gate_score_func` 重算（或保留已经按配置计算的 `gates`），并补一个从 `TopKRouter.forward` 进入的 softmax + seq-aux 回归测试；当前新增测试直接调用 `_cal_seq_aux_loss`，覆盖不到这里。

File: tests/single_card_tests/ai_edited_test/fusions/test_ai_fused_bias_swiglu_use_accuracy_compatible.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+import sys
+
+# Walk up to find the repo root.
+_test_file = os.path.abspath(__file__)
+_repo_root = _test_file
+for _ in range(10):
+    _repo_root = os.path.dirname(_repo_root)
+    if os.path.isdir(os.path.join(_repo_root, "src", "paddlefleet")):
+        break
+sys.path.insert(0, _repo_root)
```
Comment (risemeup1111) [inline tests/single_card_tests/ai_edited_test/fusions/test_ai_fused_bias_swiglu_use_accuracy_compatible.py:28]: P3 本次新增的四个测试都复制了仓库遍历、`sys.path.insert`、清空 `sys.modules` 和 standalone runner。它让测试依赖启动目录并重载全局模块状态，容易造成顺序相关行为，也命中仓库 checklist 的测试环境 hack。请由测试入口统一安装/配置工作树包，并让测试正常 import，删除四处重复 bootstrap 与 `if __name__ == "__main__"`。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [moe_layer.py:1293](https://github.com/PaddlePaddle/PaddleFleet/pull/1583#discussion_r3682937501) | P1 | ✅ |
| 2 | [moe_utils.py:223](https://github.com/PaddlePaddle/PaddleFleet/pull/1583#discussion_r3682939485) | P1 | ✅ |
| 3 | [moe_router.py:669](https://github.com/PaddlePaddle/PaddleFleet/pull/1583#discussion_r3682943854) | P1 | ✅ |
| 4 | [moe_router.py:1386](https://github.com/PaddlePaddle/PaddleFleet/pull/1583#discussion_r3682945945) | P1 | ✅ |
| 5 | [测试 bootstrap](https://github.com/PaddlePaddle/PaddleFleet/pull/1583#discussion_r3682950085) | P3 | 🟡 |
| 6 | [language_loss.py:357](https://github.com/PaddlePaddle/PaddleFleet/pull/1583#discussion_r3689585044) | P1 | 🟡 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/moe/moe_layer.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_layer.py]: 已按 #1549 的最终 head 对照。#1549 的声明范围是 GLM-4.5；其 `moe_layer.py` 也保留了这里先调用 hook、随后用 clone 覆盖结果的代码，因此之前的批准并未验证 Gemma4 组合，不能作为该组合正确的依据。当前 PR 未限制 `Gemma4MoELayer + use_accuracy_compatible=True`，所以该 P1 保持未解决；需要保留 hook 语义，或显式拒绝并测试该组合。

File: src/paddlefleet/transformer/moe/moe_utils.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_utils.py]: 对照后确认 #1549 并非没有这个问题：它曾因 padding 路径被要求修改，最终 head 在 [`moe_utils.py` L237-L263](https://github.com/PaddlePaddle/PaddleFleet/blob/4c1b46b808a8abc72d8fe27e8f636a722a2b3f50/src/paddlefleet/transformer/moe/moe_utils.py#L237-L263) 增加 `valid_token_indices`、只 reshape 有效行并 scatter 回原位置后才获批准。当前 head 仍按 `num_tokens * topk` reshape，缺少该修复，因此此 P1 保持未解决。

File: src/paddlefleet/transformer/moe/moe_router.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_router.py:669]: #1549 最终 head 确实保留了同一公式，这是此前审查的漏检，不能反向证明公式正确。Megatron-LM 的 sequence aux-loss 明确在按序列求和后执行 [`/ bsz`](https://github.com/NVIDIA/Megatron-LM/blob/ff6b92a7c3711c41f8169663614cd865233468ca/megatron/core/transformer/moe/router.py#L456-L500)，并传递有效 token 数；当前分支既未除以 `batch_size`，也未使用已计算的 `denom`，所以该 P1 保持未解决。

File: src/paddlefleet/transformer/moe/moe_router.py (hunk 4 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_router.py]: #1549 的范围和说明明确是 GLM-4.5 的 re-sigmoid 对齐；当前 PR 则是通用 MoE/MLP 改动，且新增测试配置同时设置了 `scoring_func="softmax"` 与 `use_accuracy_compatible=True`。Megatron-LM 的 aux-loss score 也会[按 `score_function` 选择 softmax/sigmoid](https://github.com/NVIDIA/Megatron-LM/blob/ff6b92a7c3711c41f8169663614cd865233468ca/megatron/core/transformer/moe/moe_utils.py#L899-L955)，因此 #1549 的 GLM sigmoid 假设不能覆盖这里的 softmax 配置；该 P1 保持未解决。

File: src/paddlefleet/transformer/moe/moe_router.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_router.py:669]: 这次提交已补上 `B>1` 的 batch mean，但 padding 归一化仍未对齐 MG。Megatron 在带 padding mask 时按有效 routing assignment 计算 `total_num_tokens`（[router.py L470-L500](https://github.com/NVIDIA/Megatron-LM/blob/ff6b92a7c3711c41f8169663614cd865233468ca/megatron/core/transformer/moe/router.py#L470-L500)；其 loss 分母使用该有效 token 数的平方，[moe_utils.py L139-L143](https://github.com/NVIDIA/Megatron-LM/blob/ff6b92a7c3711c41f8169663614cd865233468ca/megatron/core/transformer/moe/moe_utils.py#L139-L143)）。例如有效长度 `[5,3,1]`、`S=5` 时，MG 使用平均有效长度 3，而当前 `moe_router.py:652-663` 仍使用 `S²=25`。新增测试把 `loss_with_ids == loss_no_ids` 作为 expected，正好掩盖了该差异。请按有效 token 数归一化，并改为验证带 padding 的结果与 MG 公式一致。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/moe/moe_router.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_router.py:669]: 截至当前 head，仍有两项未解决：

1. P1 `src/paddlefleet/transformer/moe/moe_router.py:652-663`：本次已补上 batch mean，但带 padding 时仍以固定 `max_seq_len²` 做归一化，没有使用有效 routing token 数。Megatron-LM 在 padding mask 下使用有效 token 数计算分母；当前测试 `test_seq_aux_loss_padding_uses_sequence_length` 反而断言带/不带 `input_ids` 的结果相等，未验证参考公式。请按有效 token 数修正并补充公式对照断言。

2. P3 `tests/single_card_tests/ai_edited_test/moe/test_ai_moe_router_use_accuracy_compatible.py:20-33`：测试仍通过遍历工作树注入 `sys.path` 并清空 `sys.modules` 后再导入，依赖进程级导入顺序，测试环境中可能掩盖真实安装/导入问题。建议改用仓库既有测试启动方式，移除这段 standalone bootstrap。

已确认解决：Gemma4 hook 覆盖、混合 padding 的 permute/unpermute reshape、以及按 `gate_score_func` 保留 softmax scoring。由于 P1 仍未解决，当前不能 approve。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/moe/moe_router.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/moe/moe_router.py:669]: 本次提交已按有效 routing assignment 计算 `total_num_tokens`，并新增独立 Megatron 公式对照测试；该 seq-aux padding P1 已解决。

File: tests/single_card_tests/ai_edited_test/fusions/test_ai_fused_bias_swiglu_use_accuracy_compatible.py (hunk 5 above)
Comment (risemeup1111) [reply tests/single_card_tests/ai_edited_test/fusions/test_ai_fused_bias_swiglu_use_accuracy_compatible.py:28]: 本次只移除了 router 和 moe_utils 两个测试文件的 bootstrap；`test_ai_fused_bias_swiglu_use_accuracy_compatible.py` 与 `test_ai_three_path_clone_align_mg.py` 仍保留工作树遍历、`sys.path`/`sys.modules` 修改和 standalone runner，因此该 P3 仍未解决。

File: src/paddlefleet/models/common/language_loss/language_loss.py [hunk 6]
```diff
@@ -354,6 +354,15 @@ def _cast_loss_func(logits, labels):
                 labels, axis=1, mode=self.config.cp_balance_mode
             )
 
+        if _use_accuracy_compatible_kernel():
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/language_loss/language_loss.py:357]: P1 这里把精度开关直接当成了调试日志开关：`LanguageLoss.forward_impl` 每个 micro-batch 都会对整块 per-token loss 调用 `_md5sum()` 并 `print`，后面的新块还对标量执行 `float(loss)` 和另一次 `_md5sum()`。在 GPU/CP 训练中这会强制设备同步、持续刷屏并显著降低吞吐；而该开关是数值路径开关，不应默认承担日志输出。请移除这些临时打印，或改由已有 `LOG_LOSS_MD5`/`LOG_LAYER_MD5` 显式控制，并避免在 forward 热路径中同步。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/models/common/language_loss/language_loss.py (hunk 6 above)
Comment (zhengshengning) [reply src/paddlefleet/models/common/language_loss/language_loss.py:357]: _use_accuracy_compatible_kernel模式下需要打印出md5用于调试，所以这里没有问题

Comment (risemeup1111) [reply src/paddlefleet/models/common/language_loss/language_loss.py:357]: 已确认该开关用于精度调试，MD5 输出是该模式的预期用途；因此撤回这条性能 P1，不再要求移除这些调试打印。当前 Board 将其标记为已接受回复，剩余阻塞项仅为测试 bootstrap P3。

File: src/paddlefleet/transformer/moe/moe_router.py (hunk 3 above)
Comment (zhengshengning) [reply src/paddlefleet/transformer/moe/moe_router.py:669]: standalone bootstrap 问题作为测试，不强要求

File: tests/single_card_tests/ai_edited_test/fusions/test_ai_fused_bias_swiglu_use_accuracy_compatible.py (hunk 5 above)
Comment (risemeup1111) [reply tests/single_card_tests/ai_edited_test/fusions/test_ai_fused_bias_swiglu_use_accuracy_compatible.py:28]: 了解。该项仅属于测试实现规范建议，不作为合入要求；接受当前处理并关闭此 P3。结合前述修复与回复，当前已无未解决的代码 finding。

**Review by risemeup1111 — APPROVED** [review]

**Review by zhanghonggeng — APPROVED** [review]
